[webkitcorepy] subprocess_utils.py (run) should have the ability to stream and capture output.

bls1999 opened the pull request:
#### f95b5b76f87b07c2dafcc7c7655e7c0fdb8e6505
<pre>
[webkitcorepy] subprocess_utils.py (run) should have the ability to stream and capture output.
<a href="https://bugs.webkit.org/show_bug.cgi?id=300938">https://bugs.webkit.org/show_bug.cgi?id=300938</a>
<a href="https://rdar.apple.com/162813867">rdar://162813867</a>

Reviewed by NOBODY (OOPS!).

A deficiency of `subprocess.run` is that it can&apos;t stream output to stdout while
the process is running AND capture that output for use elsewhere. webkitcorepy
has an existing function in subprocess_utils.py (`run`) that wraps
`subprocess.run`. This change adds a new parameter to that function that
streams and captures stdout.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/subprocess_utils.py:
    (run): Add parameter `capture_and_stream_output`, change to using `Popen` instead of `run`.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f95b5b76f87b07c2dafcc7c7655e7c0fdb8e6505

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126724 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46366 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37332 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133681 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78365 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128595 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54902 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96425 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64487 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129672 "Passed tests") | [💥 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/683 "An unexpected error occured. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113339 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76950 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/126075 "Passed tests") | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31523 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77078 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31823 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136258 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53403 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104943 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/126149 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53896 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109700 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104645 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50138 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28477 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50844 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53330 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52593 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55929 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54338 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/subprocess_utils.py:
```diff
@@ -44,12 +45,20 @@ def run(*popenargs, **kwargs):
             kwargs['stdout'] = subprocess.PIPE
-            kwargs['stderr'] = subprocess.PIPE
-        return subprocess.run(*popenargs, timeout=timeout, **kwargs)
+            kwargs['stderr'] = subprocess.STDOUT
```
This is a massive change of semantics that we almost certainly cannot make without breaking existing code, and moves us away from having something API-compatible with `subprocess.run`.

bls1999 replied:
Good catch! Fixed in latest revision.

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/subprocess_utils.py:
```diff
@@ -44,12 +45,20 @@ def run(*popenargs, **kwargs):
+                sys.stdout.write(line)
+                output.append(line)
+        stdout, stderr = process.communicate(timeout=timeout)
+        return CompletedProcess(process.args, process.returncode, ''.join(output) or stdout, stderr)
```
This is entirely too simplified compared with what the stdlib `subprocess.run` does, which various kills the process in some circumstances, raises `CalledProcessError` when appropriate, etc. etc.

bls1999 replied:
This is a really good point, thank you for this feedback. Upon reflection, I think I misinterpreted some feedback @JonWBedard gave me offline during development. My most recent commit makes this much more similar to how `subprocess.run` behaves.

gsnedders reviewed (changes requested):
I don't think we can get away with changing the API like this; that we're redirecting stderr to stdout is a massive breaking change.

And you can't use such a simple approach safely with stdout and stderr separately — this is what the Python docs have a [big red scary box](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stderr) about:

> Warning Use [communicate()](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate) rather than [.stdin.write](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stdin), [.stdout.read](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stdout) or [.stderr.read](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stderr) to avoid deadlocks due to any of the other OS pipe buffers filling up and blocking the child process.

The "easiest" way to make this work correctly is to have one thread reading from stdout, one thread reading from stderr, and then the main thread waiting on it.

But I know @JonWBedard really doesn't want to use comments here, so the best bet may be to do something like create the pipes ourselves and then use `select` — though that won't work on Windows.

bls1999 committed:
[webkitcorepy] subprocess_utils.py (run) should have the ability to stream and capture output.
https://bugs.webkit.org/show_bug.cgi?id=300938
rdar://162813867

Reviewed by NOBODY (OOPS!).

A deficiency of `subprocess.run` is that it can't stream output to stdout while
the process is running AND capture that output for use elsewhere. webkitcorepy
has an existing function in subprocess_utils.py (`run`) that wraps
`subprocess.run`. This change adds a new parameter to that function that
streams and captures stdout.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/subprocess_utils.py:
    (run): Add parameter `capture_and_stream_output`, change to using `Popen` instead of `run`.
